### ISSUE TITLE
logstash can't read data from redis sometimes

### ISSUE THREAD

**Issue by GodHong** [logstash-plugins/logstash-input-redis#13]
It works most of time,but will not work sometimes, and the reason is that logstash can't read data from redis anymore. My logstash version is 1.4.2.
input {
redis {
host => localhost
data_type => "list"
key => "paylist"
type => "pay"
}
redis {
host => localhost
data_type => "list"
key => "compensationlist"
type => "compensation"
}
redis {
host => localhost
data_type => "list"
key => "orderlist"
type => "order"
}
redis {
host => localhost
data_type => "list"
key => "list4log"
type => "log"
}
}

**Comment by GodHong** [comment]
restart logstash with conf, it will work again. Thanks!

**Comment by GodHong** [comment]
I have a http output, is this the reason? and the number of log is large!

**Comment by guyboertje** [comment]
@GodHong - have you got any more information? For example - how do you know that logstash cannot read from redis at a certain point? Is it because you see something in the logger output or that you stop seeing stuff in the http output.

**Comment by guyboertje** [comment]
@GodHong - please post you full logstash config

**Comment by GodHong** [comment]
@guyboertje - I know that by redis-cli , the list length increase always, but never decrease.
full logstash config as follow:
input {
redis {
host => "127.0.0.1"
data_type => "list"
key => "paylist"
type => "pay"
}

redis {
host => "127.0.0.1"
data_type => "list"
key => "compensationlist"
type => "compensation"
}

redis {
host => "127.0.0.1"
data_type => "list"
key => "orderlist"
type => "order"
}

redis {
host => "127.0.0.1"
data_type => "list"
key => "list4log"
type => "log"
}
}

output {
redis {
host => "127.0.0.1"

data_type =>"list"
key => "logstash:redis"
}

http{
url => "*****"
http_method => "post"
format => "json"
headers => ["Token", "*****"]
workers => 8
}

}

**Comment by GodHong** [comment]
@guyboertje - BTW it works most of time, but will not work sometimes, need restart loogstash.

**Comment by GodHong** [comment]
@guyboertje - there's not http output before and it works all the time, Did  http output  cause problem?

**Comment by guyboertje** [comment]
@Godhong - well they both use networking.  I'm going to investigate.

**Comment by GodHong** [comment]
@guyboertje - thanks!

**Comment by guyboertje** [comment]
@GodHong - have you found the answer independently? Based on my experimentation, your pipeline is stalling.  The http output can't keep pace with your 4 redis inputs.  The events are dispatched to your outputs sequentially in a loop so essentially the internal queue is only drained as fast as your slowest output, and the actual speed of the http output is subject to network latencies and therefore not deterministic. What webserver are you pointing the http output at?

**Comment by slaupster** [comment]
I have this problem too @ logstash 2.1.1 and it seems to be related to load of the server when logstash starts; it doesn't work if the server is busy but if its not busy its fine.  This is redis -> elasticsearch.  

It feels like Redis input stalls, but as @guyboertje suggests, it *could* be related to elasticsearch not being fully ready to accept work but there are no errors in the logs, and it never actually recovers until a restart (not clear whether restarting is kicking elasticsearch output or redis input).

Javacore indicates that both redis and elasticsearch are in a waiting state but obviously the interceding ruby makes it difficult to understand which one is waiting erroneously.

How do we get diagnostics for this ?  I can reliably recreate.

**Comment by slaupster** [comment]
Redis thread is at:

3XMTHREADINFO      "<redis" J9VMThread:0x0000000001B35600, j9thread_t:0x00007FED08502EB0, java/lang/Thread:0x00000000E332C6C0, state:CW, prio=5
3XMJAVALTHREAD            (java/lang/Thread getId:0x16, isDaemon:true)
3XMTHREADINFO1            (native thread ID:0x56, native priority:0x5, native policy:UNKNOWN, vmstate:CW, vm thread flags:0x00000101)
3XMTHREADINFO2            (native stack address range from:0x00007FED503CE000, to:0x00007FED5040F000, size:0x41000)
3XMCPUTIME               CPU usage total: 0.630409070 secs, current category="Application"
3XMTHREADBLOCK     Waiting on: org/jruby/ext/thread/SizedQueue@0x00000000E3330348 Owned by: <unowned>
3XMHEAPALLOC             Heap bytes allocated since last GC cycle=0 (0x0)
3XMTHREADINFO3           Java callstack:
4XESTACKTRACE                at java/lang/Object.wait(Native Method)
4XESTACKTRACE                at org/jruby/RubyThread$SleepTask.run(RubyThread.java:1049)
5XESTACKTRACE                   (entered lock: org/jruby/ext/thread/SizedQueue@0x00000000E3330348, entry count: 2)
4XESTACKTRACE                at org/jruby/RubyThread.executeBlockingTask(RubyThread.java:1065(Compiled Code))
4XESTACKTRACE                at org/jruby/RubyThread.wait_timeout(RubyThread.java:1413)
4XESTACKTRACE                at org/jruby/ext/thread/SizedQueue.push(SizedQueue.java:133)
5XESTACKTRACE                   (entered lock: org/jruby/ext/thread/SizedQueue@0x00000000E3330348, entry count: 1)

**Comment by guyboertje** [comment]
@slaupster - This `3XMTHREADBLOCK Waiting on: org/jruby/ext/thread/SizedQueue@0x00000000E3330348 Owned by: ` confirms that the redis input thread is blocked waiting on the Ruby SizedQueue to empty a bit.

What is your config?

**Comment by slaupster** [comment]
hey @guyboertje thanks for the reply.  Config:

input {
	redis {
		host => "%redis_host%"
		port => "%redis_port%"
		data_type => "list"
		key => "logstash"
	}
}
filter.....
elasticsearch {
			host => "%es_host%"
			port => "%es_host%"
			cluster => "%es_cluster_name%"
			protocol => "transport"
			index => "%es_cluster_name%-debug-%{+YYYY.MM.dd}"
			template => "/etc/es-%es_cluster_name%_template.json"
			template_name => "%es_cluster_name%"
		}

We sort out the %bla% at docker start.  es and redis both run locally always.  The times when we see the problem are always when the node is under heavy load.

There are many elasticsearch threads, mostly parked or in netty select().  There are some in ruby waits().  If you feel that stack is indicating es output is stalled, should this head over to that plugin ?

Thanks again!

**Comment by slaupster** [comment]
PS we also have gates to ensure the hosts are resolvable before starting logstash.

**Comment by slaupster** [comment]
hi @guyboertje - please can you suggest the next step here ?  Move to elasticsearch output ?

thanks!

**Comment by guyboertje** [comment]
@slaupster - I think using the ES output is the way to go.

**Comment by slaupster** [comment]
@guyboertje we are already using ES output.  Sorry I think I didn't explain myself well enough - I meant do you think this issue is an ES output issue rather than a Redis input issue ?  If so I should open an issue there ?

**Comment by guyboertje** [comment]
@slaupster - from what you are saying, the problem is hard to reproduce.  There is probably not much point in creating an issue in the ES output repo unless you have a concrete illustration of the exact fault.

Can I suggest that you use [our discussion forum](https://discuss.elastic.co/)?
